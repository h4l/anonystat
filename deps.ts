export { getReasonPhrase, StatusCodes } from "npm:http-status-codes@2";
export { z } from "https://deno.land/x/zod@v3.22.4/mod.ts";
export {
  assert,
  unreachable,
} from "https://deno.land/std@0.221.0/assert/mod.ts";
export {
  difference,
} from "https://deno.land/std@0.221.0/datetime/difference.ts";
export type {
  Unit,
} from "https://deno.land/std@0.221.0/datetime/difference.ts";

export { parse as parseJsonc } from "https://deno.land/std@0.221.0/jsonc/mod.ts";
export type { JsonValue } from "https://deno.land/std@0.221.0/jsonc/mod.ts";
export { generate } from "https://deno.land/std@0.221.0/uuid/v5.ts";
export type { Closer } from "https://deno.land/std@0.221.0/io/types.ts";
