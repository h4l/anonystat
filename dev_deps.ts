export {
  assert,
  assertAlmostEquals,
  assertEquals,
  assertGreater,
  assertInstanceOf,
  AssertionError,
  assertLess,
  assertLessOrEqual,
  assertMatch,
  assertNotEquals,
  assertRejects,
  assertStringIncludes,
} from "https://deno.land/std@0.221.0/assert/mod.ts";

export { assertSnapshot } from "https://deno.land/std@0.221.0/testing/snapshot.ts";

export {
  assertSpyCall,
  spy,
  stub,
} from "https://deno.land/std@0.221.0/testing/mock.ts";

export type { Stub } from "https://deno.land/std@0.221.0/testing/mock.ts";

export { FakeTime } from "https://deno.land/std@0.221.0/testing/time.ts";

import { validate } from "https://deno.land/std@0.221.0/uuid/mod.ts";
export const validateUuid = validate;

export { toText } from "https://deno.land/std@0.221.0/streams/mod.ts";

export { fromFileUrl } from "https://deno.land/std@0.221.0/path/posix/from_file_url.ts";
