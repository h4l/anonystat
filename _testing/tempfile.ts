import { Closer } from "../deps.ts";

/** A reference to a temporary file that's deleted when disposed.
 *
 * ```ts
 * using file = await Deno.open("/foo/bar.txt", { read: true });
 * const fileInfo = await file.stat();
 * if (fileInfo.isFile) {
 *   const buf = new Uint8Array(100);
 *   const numberOfBytesRead = await file.read(buf); // 11 bytes
 *   const text = new TextDecoder().decode(buf);  // "hello world"
 * }
 * ```
 */
export class TemporaryFile implements Closer, Disposable {
  constructor(readonly path: string) {}

  /** Create a TemporaryFile, like {@linkcode Deno.makeTempFile}. */
  static async create(
    options: Deno.MakeTempOptions = {},
  ): Promise<TemporaryFile> {
    return new TemporaryFile(await Deno.makeTempFile(options));
  }

  [Symbol.dispose](): void {
    this.close();
  }

  #closed: boolean = false;
  close(): void {
    if (this.#closed) return;
    this.#closed = true;
    Deno.remove(this.path).catch(() => {});
  }
}
