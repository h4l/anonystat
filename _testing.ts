import { string } from "https://deno.land/x/zod@v3.22.4/types.ts";
import { ErrorResult, Result, SuccessResult } from "./_misc.ts";
import {
  assert,
  assertEquals,
  AssertionError,
  assertLess,
  assertLessOrEqual,
  Stub,
  validateUuid,
} from "./dev_deps.ts";
import { assertFalse } from "https://deno.land/std@0.221.0/assert/assert_false.ts";
import { AnyPayload } from "./payload-schemas.ts";
import { assertGreater } from "./dev_deps.ts";

/** Parse an iso date and return it as a ms timestamp, throw on invalid syntax. */
export function timestamp(isoDate: string): number {
  const timestamp = Date.parse(isoDate);
  if (Number.isNaN(timestamp)) throw new Error("Invalid date");
  return timestamp;
}

/** Parse an iso date and return it as a Date, throw on invalid syntax. */
export function date(isoDate: string, offset: number = 0): Date {
  return new Date(timestamp(isoDate) + offset);
}

export function assertSuccessful<T, E>(
  result: Result<T, E>,
): asserts result is SuccessResult<T> {
  if (result.success) return;
  assert(
    false,
    `expected result to be successful but it failed with error: ${
      Deno.inspect(result.error)
    }`,
  );
}

export function assertUnsuccessful<T, E>(
  result: Result<T, E>,
): asserts result is ErrorResult<E> {
  if (!result.success) return;
  assert(
    false,
    `result is not unsuccessful with data: ${Deno.inspect(result.data)}`,
  );
}

/** Assert that value is a valid UUID string. */
export function assertUuid(value: unknown): asserts value is string {
  if (typeof value === "string" && validateUuid(value)) return;
  assert(false, `value is not a UUID: ${Deno.inspect(value)}`);
}

/** Linearly interpolate from `a` to `b` by `t`. */
export function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

export function assertResponseOk(response: Response): void {
  if (response.status >= 200 && response.status < 300) return;
  assert(
    false,
    `expected response to have 2XX status but is ${response.status}: ${
      Deno.inspect(response)
    }`,
  );
}

export async function assertOneFetchForMpCollectRequest(
  fetch: Stub<
    typeof globalThis,
    [input: string | Request | URL, init?: RequestInit | undefined],
    Promise<Response>
  >,
  mpCollectRequestAttrs: MpCollectRequestAttrs,
) {
  await assertFetchForMpCollectRequest(
    fetch,
    { index: 0, maxCalls: 1 },
    mpCollectRequestAttrs,
  );
}
export async function assertFetchForMpCollectRequest(
  fetch: Stub<
    typeof globalThis,
    [input: string | Request | URL, init?: RequestInit | undefined],
    Promise<Response>
  >,
  { index, maxCalls }: { index: number; maxCalls?: number },
  mpCollectRequestAttrs: MpCollectRequestAttrs,
) {
  if (maxCalls !== undefined) {
    assertLessOrEqual(fetch.calls.length, maxCalls, "Too many fetch() calls");
  }
  assertGreater(fetch.calls.length, index, "Too few fetch() calls");

  const request = fetch.calls[index].args[0];
  if (
    !(fetch.calls[index].args.length === 1 && request instanceof Request)
  ) {
    assert(
      false,
      `expected fetch to be called with one Request argument: ${
        Deno.inspect(fetch.calls[index].args)
      }`,
    );
  }
  await assertIsMpRequest(request, mpCollectRequestAttrs);
}

export type MpCollectRequestAttrs = {
  origin: string;
  measurement_id: string;
  api_secret: string;
  user_id?: string | ((value: string | undefined) => void);
};

export async function assertIsMpRequest(
  request: Request,
  { origin, measurement_id, api_secret, user_id }: MpCollectRequestAttrs,
): Promise<void> {
  const reqUrl = new URL(request.url);
  assertEquals(reqUrl.origin, origin);
  assertEquals(reqUrl.pathname, "/mp/collect");
  assertEquals(reqUrl.searchParams.get("measurement_id"), measurement_id);
  assertEquals(reqUrl.searchParams.get("api_secret"), api_secret);

  if (user_id !== undefined) {
    const payload = await getRequestPayload(request);
    if (typeof user_id === "string") {
      assertEquals(
        payload.user_id,
        user_id,
        "Payload contains unexpected user_id",
      );
    } else {
      user_id(payload.user_id);
    }
  }
}

async function getRequestPayload(request: Request): Promise<AnyPayload> {
  try {
    return AnyPayload.parse(await request.json());
  } catch (e) {
    throw new AssertionError(
      `Request body is not readable as a collect request payload: ${
        Deno.inspect(e)
      }`,
    );
  }
}
