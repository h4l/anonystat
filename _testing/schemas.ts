import { assert, assertEquals } from "../dev_deps.ts";

import { z } from "../deps.ts";

export function repeat<T>(value: T, times: number): T[] {
  const values = new Array(times);
  values.fill(value);
  return values;
}

export function repeatString(
  value: string,
  times: number,
  joiner: string = "",
): string {
  return repeat(value, times).join(joiner);
}

export const Input = Symbol("Input");

function getInput<Schema extends z.ZodTypeAny>(options: InputOptions<Schema>) {
  return "input" in options ? options.input : options.rawInput;
}

type InputOptions<Schema extends z.ZodTypeAny> = {
  input: z.input<Schema>;
} | { rawInput: unknown };

type MultipleInputOptions<Schema extends z.ZodTypeAny> = {
  inputs: ReadonlyArray<z.input<Schema>>;
} | { rawInputs: ReadonlyArray<unknown> };

function getInputs<Schema extends z.ZodTypeAny>(
  options: MultipleInputOptions<Schema>,
): ReadonlyArray<InputOptions<Schema>> {
  return "inputs" in options
    ? options.inputs.map((i) => ({ input: i }))
    : options.rawInputs.map((ri) => ({ rawInput: ri }));
}

export function assertParses<Schema extends z.ZodTypeAny>(
  { schema, expected = Input, ...inputs }: {
    schema: Schema;
    expected?: z.infer<Schema> | typeof Input;
  } & InputOptions<Schema>,
): void {
  const input = getInput(inputs);
  const result = schema.safeParse(input);
  if (!result.success) {
    assert(
      false,
      `Schema did not parse value: ${
        Deno.inspect({ input, errors: result.error.format()._errors })
      }`,
    );
  }
  assertEquals(result.data, expected === Input ? input : expected);
}

export function assertRejects<Schema extends z.ZodTypeAny>(
  { schema, ...inputs }: {
    schema: Schema;
  } & InputOptions<Schema>,
): void {
  const input = getInput(inputs);
  const result = schema.safeParse(input);
  if (result.success) {
    assert(
      false,
      `Schema parse unexpectedly succeeded: ${
        Deno.inspect({ input, result: result.data })
      }`,
    );
  }
}

export class SchemaAssertions<Schema extends z.ZodTypeAny> {
  constructor(readonly schema: Schema) {}

  assertParses(
    input: z.input<Schema>,
    expected: z.infer<Schema> | typeof Input = Input,
  ): void {
    assertParses({ schema: this.schema, input, expected });
  }

  assertParsesAll(
    options: MultipleInputOptions<Schema>,
  ): void {
    for (const inputOptions of getInputs(options)) {
      assertParses({ schema: this.schema, ...inputOptions });
    }
  }

  assertRejects(
    input: z.input<Schema>,
  ): void {
    assertRejects({ schema: this.schema, input });
  }

  assertRejectsAll(
    options: MultipleInputOptions<Schema>,
  ): void {
    for (const inputOptions of getInputs(options)) {
      assertRejects({ schema: this.schema, ...inputOptions });
    }
  }
}
