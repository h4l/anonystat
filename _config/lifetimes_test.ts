import { assertSuccessful, assertUnsuccessful, date } from "../_testing.ts";
import { z } from "../deps.ts";
import { assert, assertEquals, assertSnapshot } from "../dev_deps.ts";

import {
  EvaluatedDisambiguatedLifetimeExpression,
  formatCompactIsoDateTime,
  formatIsoInterval,
  ParsedIsoIntervalLifetime,
  ValidatedDisambiguatedLifetimeExpression,
} from "./lifetimes.ts";

Deno.test("IsoIntervalLifetime", async (t) => {
  const IsoIntervalLifetime = ParsedIsoIntervalLifetime.transform((arg) =>
    arg.lifetime
  );
  await t.step("optional infinite recurring prefix", () => {
    [
      "R/2001-02-03T04:05:06Z/P1Y",
      "r/2001-02-03T04:05:06Z/P1Y",
      "2001-02-03 04:05:06Z/P1Y",
    ].forEach((interval) => {
      assertEquals(
        IsoIntervalLifetime.parse(interval),
        { unit: "years", count: 1, from: date("2001-02-03T04:05:06Z") },
      );
    });
  });

  await t.step("from milliseconds", () => {
    assertEquals(
      IsoIntervalLifetime.parse("R/2001-02-03T04:05:06.45Z/P1Y"),
      { unit: "years", count: 1, from: date("2001-02-03T04:05:06.45Z") },
    );
  });

  await t.step("from without time", () => {
    assertEquals(
      IsoIntervalLifetime.parse("R/2001-02-03/P1Y"),
      { unit: "years", count: 1, from: date("2001-02-03T00:00:00Z") },
    );
  });

  await t.step("units & durations", () => {
    assertEquals(
      IsoIntervalLifetime.parse("R/2001-02-03T04:05:06Z/P2Y"),
      { unit: "years", count: 2, from: date("2001-02-03T04:05:06Z") },
    );

    assertEquals(
      IsoIntervalLifetime.parse("R/2001-02-03T04:05:06Z/P2M"),
      { unit: "months", count: 2, from: date("2001-02-03T04:05:06Z") },
    );

    assertEquals(
      IsoIntervalLifetime.parse("R/2001-02-03T04:05:06Z/P12W"),
      { unit: "weeks", count: 12, from: date("2001-02-03T04:05:06Z") },
    );

    assertEquals(
      IsoIntervalLifetime.parse("R/2001-02-03T04:05:06Z/P100D"),
      { unit: "days", count: 100, from: date("2001-02-03T04:05:06Z") },
    );

    assertEquals(
      IsoIntervalLifetime.parse("R/2001-02-03T04:05:06Z/PT2H"),
      { unit: "hours", count: 2, from: date("2001-02-03T04:05:06Z") },
    );
  });

  await t.step("multiple periods not supported", () => {
    assert(
      !IsoIntervalLifetime.safeParse("R/2001-02-03T04:05:06Z/P2M1W").success,
    );
    assert(
      !IsoIntervalLifetime.safeParse("R/2001-02-03T04:05:06Z/P2MT1H").success,
    );
  });
});

function joinedErrorMessages(error: z.ZodError): string {
  return error.issues.map((i) => i.message).join("; ");
}

Deno.test("ValidatedDisambiguatedLifetimeExpression", async (t) => {
  await t.step("accepts valid input", () => {
    assertSuccessful(
      ValidatedDisambiguatedLifetimeExpression.safeParse("1 month"),
    );
    assertSuccessful(
      ValidatedDisambiguatedLifetimeExpression.safeParse("R/2024-01-01/P1M"),
    );
  });

  await t.step(
    "rejects invalid simple period with appropriate message",
    async (t) => {
      const result = ValidatedDisambiguatedLifetimeExpression.safeParse(
        "3 seconds",
      );
      assertUnsuccessful(result);
      await assertSnapshot(t, joinedErrorMessages(result.error));
    },
  );

  await t.step(
    "rejects invalid non-simple period with appropriate message",
    async (t) => {
      const result = ValidatedDisambiguatedLifetimeExpression.safeParse(
        "abc/123",
      );
      assertUnsuccessful(result);
      await assertSnapshot(t, joinedErrorMessages(result.error));
    },
  );
});

Deno.test("formatCompactIsoDateTime()", () => {
  assertEquals(
    formatCompactIsoDateTime(date("2024-02-03T00:00:00Z")),
    "2024-02-03",
  );
  assertEquals(
    formatCompactIsoDateTime(date("2024-02-03T00:00:00.500Z")),
    "2024-02-03T00:00:00.500Z",
  );
});

Deno.test("formatIsoInterval()", () => {
  assertEquals(
    formatIsoInterval({ from: 0, unit: "hours", count: 2 }),
    "R/1970-01-01/PT2H",
  );
  assertEquals(
    formatIsoInterval({ unit: "hours", count: 2 }),
    "R/1970-01-01/PT2H",
  );
  assertEquals(
    formatIsoInterval({
      from: date("2024-02-03T04:05:06.500Z"),
      unit: "weeks",
      count: 2,
    }),
    "R/2024-02-03T04:05:06.500Z/P2W",
  );
});
