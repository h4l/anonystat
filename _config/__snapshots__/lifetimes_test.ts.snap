export const snapshot = {};

snapshot[`ValidatedDisambiguatedLifetimeExpression > rejects invalid simple period with appropriate message 1`] = `\`Value must be "[<number>] <unit>" where number is 1+ and unit is one of seconds, minutes, hours, days, weeks, months, quarters, years (ignoring case, with or without 's'); Value must be an interval like "2 weeks" or an ISO interval with a start time, like "R/2024-01-01/P2W"\``;

snapshot[`ValidatedDisambiguatedLifetimeExpression > rejects invalid non-simple period with appropriate message 1`] = `'Value is not an ISO 8601 interval with a single period; Value must be an interval like "2 weeks" or an ISO interval with a start time, like "R/2024-01-01/P2W"'`;
